# ETL Completo

In [109]:
# Importación de librerías
import pandas as  pd
import numpy as np

## products

In [110]:
# Ingestamos dataset
url = 'https://raw.githubusercontent.com/ronalcabrera/PG_Olist/main/Datasets/olist_products_dataset.csv'
products = pd.read_csv(url, delimiter=',', encoding='UTF-8')

In [111]:
# Obervamos las primeras 15 filas del dataset
products.head(15)

product_id             product_category_name  \
0   1e9e8ef04dbcff4541ed26657ea517e5                        perfumaria   
1   3aa071139cb16b67ca9e5dea641aaa2f                             artes   
2   96bd76ec8810374ed1b65e291975717f                     esporte_lazer   
3   cef67bcfe19066a932b7673e239eb23d                             bebes   
4   9dc1a7de274444849c219cff195d0b71             utilidades_domesticas   
5   41d3672d4792049fa1779bb35283ed13             instrumentos_musicais   
6   732bd381ad09e530fe0a5f457d81becb                        cool_stuff   
7   2548af3e6e77a690cf3eb6368e9ab61e                  moveis_decoracao   
8   37cc742be07708b53a98702e77a21a02                  eletrodomesticos   
9   8c92109888e8cdf9d66dc7e463025574                        brinquedos   
10  14aa47b7fe5c25522b47b4b29c98dcb9                   cama_mesa_banho   
11  03b63c5fc16691530586ae020c345514                             bebes   
12  cf55509ea8edaaac1d28fdb16e48fc22             instrumentos_musicais   
13  7bb6f29c2be57716194f96496660c7c2                  moveis_decoracao   
14  eb31436580a610f202c859463d8c7415  construcao_ferramentas_seguranca   

    product_name_lenght  product_description_lenght  product_photos_qty  \
0                  40.0                       287.0                 1.0   
1                  44.0                       276.0                 1.0   
2                  46.0                       250.0                 1.0   
3                  27.0                       261.0                 1.0   
4                  37.0                       402.0                 4.0   
5                  60.0                       745.0                 1.0   
6                  56.0                      1272.0                 4.0   
7                  56.0                       184.0                 2.0   
8                  57.0                       163.0                 1.0   
9                  36.0                      1156.0                 1.0   
10                 54.0                       630.0                 1.0   
11                 49.0                       728.0                 4.0   
12                 43.0                      1827.0                 3.0   
13                 51.0                      2083.0                 2.0   
14                 59.0                      1602.0                 4.0   

    product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0              225.0               16.0               10.0              14.0  
1             1000.0               30.0               18.0              20.0  
2              154.0               18.0                9.0              15.0  
3              371.0               26.0                4.0              26.0  
4              625.0               20.0               17.0              13.0  
5              200.0               38.0                5.0              11.0  
6            18350.0               70.0               24.0              44.0  
7              900.0               40.0                8.0              40.0  
8              400.0               27.0               13.0              17.0  
9              600.0               17.0               10.0              12.0  
10            1100.0               16.0               10.0              16.0  
11            7150.0               50.0               19.0              45.0  
12             250.0               17.0                7.0              17.0  
13             600.0               68.0               11.0              13.0  
14             200.0               17.0                7.0              17.0

products:
- Probablemente lo más adecuado sería traducir las categorías para normalizar el idioma

Analizamos nombres de columnas, cantidades y tipos de datos

products.info()

products:
- Hay valores nulos en algunas columnas.
- Los tipos de datos parecen ser acordes. Aunque habría que ver el tema del id

In [112]:
categorias = products.groupby(['product_category_name'])['product_category_name'].count()
categorias

product_category_name
agro_industria_e_comercio      74
alimentos                      82
alimentos_bebidas             104
artes                          55
artes_e_artesanato             19
                             ... 
sinalizacao_e_seguranca        93
tablets_impressao_imagem        9
telefonia                    1134
telefonia_fixa                116
utilidades_domesticas        2335
Name: product_category_name, Length: 73, dtype: int64

In [113]:
# Ingesto el dataset de categorias en inglés, para comparar
url = 'https://raw.githubusercontent.com/ronalcabrera/PG_Olist/main/Datasets/product_category_name_translation.csv'
cat_ingles = pd.read_csv(url, delimiter=',', encoding='UTF-8')

In [114]:
cat_ingles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


Observamos que 'cat_ingles' tiene dos categorías menos. Estas categorías son:

- portateis_cozinha_e_preparadores_de_alimentos = kitchen_and_food_preparation_racks
- pc_gamer = pc_gamer


In [115]:
# Vamos a proceder a agregar estas categorías:
cat_ingles.loc[len(cat_ingles)] = ['portateis_cozinha_e_preparadores_de_alimentos', 'kitchen_and_food_preparation_racks']
cat_ingles.loc[len(cat_ingles)] = ['pc_gamer', 'pc_gamer']

cat_ingles

In [116]:
# Quitamos los guiones
cat_ingles['product_category_name_english'] = cat_ingles['product_category_name_english'].str.replace('_',' ')
# Ponemos la primer letra en mayúscula
cat_ingles['product_category_name_english'] = cat_ingles['product_category_name_english'].str.capitalize()

In [117]:
# Reemplazamos los nombres en portugués por los nombres en inglés:
products['product_category_name'] = products['product_category_name'].map(cat_ingles.set_index('product_category_name')['product_category_name_english'])

In [118]:
products

product_id      product_category_name  \
0      1e9e8ef04dbcff4541ed26657ea517e5                  Perfumery   
1      3aa071139cb16b67ca9e5dea641aaa2f                        Art   
2      96bd76ec8810374ed1b65e291975717f             Sports leisure   
3      cef67bcfe19066a932b7673e239eb23d                       Baby   
4      9dc1a7de274444849c219cff195d0b71                 Housewares   
...                                 ...                        ...   
32946  a0b7d5a992ccda646f2d34e418fff5a0            Furniture decor   
32947  bf4538d88321d0fd4412a93c974510e6  Construction tools lights   
32948  9a7c6041fa9592d9d9ef6cfe62a71f8c             Bed bath table   
32949  83808703fc0706a22e264b9d75f04a2e      Computers accessories   
32950  106392145fca363410d287a815be6de4             Bed bath table   

       product_name_lenght  product_description_lenght  product_photos_qty  \
0                     40.0                       287.0                 1.0   
1                     44.0                       276.0                 1.0   
2                     46.0                       250.0                 1.0   
3                     27.0                       261.0                 1.0   
4                     37.0                       402.0                 4.0   
...                    ...                         ...                 ...   
32946                 45.0                        67.0                 2.0   
32947                 41.0                       971.0                 1.0   
32948                 50.0                       799.0                 1.0   
32949                 60.0                       156.0                 2.0   
32950                 58.0                       309.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
0                 225.0               16.0               10.0   
1                1000.0               30.0               18.0   
2                 154.0               18.0                9.0   
3                 371.0               26.0                4.0   
4                 625.0               20.0               17.0   
...                 ...                ...                ...   
32946           12300.0               40.0               40.0   
32947            1700.0               16.0               19.0   
32948            1400.0               27.0                7.0   
32949             700.0               31.0               13.0   
32950            2083.0               12.0                2.0   

       product_width_cm  
0                  14.0  
1                  20.0  
2                  15.0  
3                  26.0  
4                  13.0  
...                 ...  
32946              40.0  
32947              16.0  
32948              27.0  
32949              20.0  
32950               7.0  

[32951 rows x 9 columns]

In [119]:
# Búsqueda de valores únicos
products.nunique()

product_id                    32951
product_category_name            73
product_name_lenght              66
product_description_lenght     2960
product_photos_qty               19
product_weight_g               2204
product_length_cm                99
product_height_cm               102
product_width_cm                 95
dtype: int64

Hay 610 registros en product_category_name que están vacíos. Adicionalmente no cuentan con datos en otras columnas. Como son pocos registros comparados con el total, y como no disponemos de forma de identificarlos, procedemos a eliminarlos.

In [120]:
products = products[products.product_category_name.notnull()] # elimina 610 registros
products = products[products.product_weight_g.notnull()] # elimina 1 registro

In [121]:
products.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32340 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32340 non-null  object 
 1   product_category_name       32340 non-null  object 
 2   product_name_lenght         32340 non-null  float64
 3   product_description_lenght  32340 non-null  float64
 4   product_photos_qty          32340 non-null  float64
 5   product_weight_g            32340 non-null  float64
 6   product_length_cm           32340 non-null  float64
 7   product_height_cm           32340 non-null  float64
 8   product_width_cm            32340 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.5+ MB


In [122]:
# Exportamos dataset limpio
products.to_csv('./Datasets_procesados/products.csv', index=False)

## sellers

In [123]:
# Ingestamos dataset
url = 'https://raw.githubusercontent.com/ronalcabrera/PG_Olist/main/Datasets/olist_sellers_dataset.csv'
sellers = pd.read_csv(url, delimiter=',', encoding='UTF-8')
sellers

seller_id  seller_zip_code_prefix  \
0     3442f8959a84dea7ee197c632cb2df15                   13023   
1     d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2     ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3     c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4     51a04a8a6bdcb23deccc82b0b80742cf                   12914   
...                                ...                     ...   
3090  98dddbc4601dd4443ca174359b237166                   87111   
3091  f8201cab383e484733266d1906e2fdfa                   88137   
3092  74871d19219c7d518d0090283e03c137                    4650   
3093  e603cf3fec55f8697c9059638d6c8eb5                   96080   
3094  9e25199f6ef7e7c347120ff175652c3b                   12051   

            seller_city seller_state  
0              campinas           SP  
1            mogi guacu           SP  
2        rio de janeiro           RJ  
3             sao paulo           SP  
4     braganca paulista           SP  
...                 ...          ...  
3090            sarandi           PR  
3091            palhoca           SC  
3092          sao paulo           SP  
3093            pelotas           RS  
3094            taubate           SP  

[3095 rows x 4 columns]

### Columnas `seller_id`

In [124]:
# Verifico si hay duplicados:
sellers[['seller_id','seller_city']].groupby(['seller_id']).count()

seller_city
seller_id                                    
0015a82c2db000af6aaaf3ae2ecb0532            1
001cca7ae9ae17fb1caed9dfb1094831            1
001e6ad469a905060d959994f1b41e4f            1
002100f778ceb8431b7a1020ff7ab48f            1
003554e2dce176b5555353e4f3555ac8            1
...                                       ...
ffcfefa19b08742c5d315f2791395ee5            1
ffdd9f82b9a447f6f8d4b91554cc7dd3            1
ffeee66ac5d5a62fe688b9d26f83f534            1
fffd5413c0700ac820c7069d66d98c89            1
ffff564a4f9085cd26170f4732393726            1

[3095 rows x 1 columns]

Se verifica que hay la misma cantidad de filas luego de agrupar por `seller_id`, no hay seller_id duplicados.

### Columna `seller_zip_code_prefix`

In [125]:
# Verifico el tipo de dato este correcto
sellers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [126]:
# Por el tipo de dato de la columna `customer_zip_code_prefix` se verifica que son todos números enteros, procedo a listarlos y contarlos
zip_code_prefix = sellers['seller_zip_code_prefix'].unique().tolist()
zip_code_prefix.sort()
len(zip_code_prefix)

2246

Verifico que los prefijos y las abreviaturas esten bien

In [127]:
# Primero verifico que los valores mínimos esten dentro del rango, luego verifico los máximos
df = sellers[['seller_state','seller_zip_code_prefix']].groupby(['seller_state'], ).min()
df['prefix_max'] = sellers[['seller_state','seller_zip_code_prefix']].groupby(['seller_state'], ).max().values
df.sort_values('seller_zip_code_prefix')

seller_zip_code_prefix  prefix_max
seller_state                                    
SP                              1001       95076
RJ                             20020       28990
RN                             21210       59775
ES                             29010       29704
MG                             30111       39801
BA                             40130       48602
SE                             49055       49980
PE                             50751       55602
PB                             58030       58865
CE                             60025       63540
PI                             64033       64033
MA                             65072       65072
AM                             69005       69005
AC                             69900       69900
DF                             70070       73020
GO                             72801       76500
RO                             76804       76900
MT                             78020       78820
MS                             79090       79400
PR                             80010       87900
PA                             85960       85960
SC                             88010       89900
RS                             90010       99730

### Columna `seller_city`
No vamos a trabajar con esta columna, asi que no se le hará ningún tratamiento

### Columna `seller_state`

In [128]:
print('Cantidad de federaciones:',len(sellers.seller_state.unique().tolist()))
print(sellers.seller_state.unique().tolist())

Cantidad de federaciones: 23
['SP', 'RJ', 'PE', 'PR', 'GO', 'SC', 'BA', 'DF', 'RS', 'MG', 'RN', 'MT', 'CE', 'PB', 'AC', 'ES', 'RO', 'PI', 'MS', 'SE', 'MA', 'AM', 'PA']


In [129]:
# Exportamos dataset limpio
products.to_csv('./Datasets_procesados/sellers.csv', index=False)

## customers

In [130]:
# Ingestamos dataset
url = 'https://raw.githubusercontent.com/ronalcabrera/PG_Olist/main/Datasets/olist_customers_dataset.csv'
customers = pd.read_csv(url, delimiter=',', encoding='UTF-8')
customers

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code_prefix          customer_city customer_state  
0                         14409                 franca             SP  
1                          9790  sao bernardo do campo             SP  
2                          1151              sao paulo             SP  
3                          8775        mogi das cruzes             SP  
4                         13056               campinas             SP  
...                         ...                    ...            ...  
99436                      3937              sao paulo             SP  
99437                      6764        taboao da serra             SP  
99438                     60115              fortaleza             CE  
99439                     92120                 canoas             RS  
99440                      6703                  cotia             SP  

[99441 rows x 5 columns]

### Columnas `customer_id` & `customer_unique_id`

In [131]:
# Verifico si hay duplicados:
customers[['customer_id','customer_unique_id']].groupby(['customer_id'])['customer_unique_id'].count()

customer_id
00012a2ce6f8dcda20d059ce98491703    1
000161a058600d5901f007fab4c27140    1
0001fd6190edaaf884bcaf3d49edf079    1
0002414f95344307404f0ace7a26f1d5    1
000379cdec625522490c315e70c7a9fb    1
                                   ..
fffecc9f79fd8c764f843e9951b11341    1
fffeda5b6d849fbd39689bb92087f431    1
ffff42319e9b2d713724ae527742af25    1
ffffa3172527f765de70084a7e53aae8    1
ffffe8b65bbe3087b653a978c870db99    1
Name: customer_unique_id, Length: 99441, dtype: int64

Se verifica que hay la misma cantidad de filas luego de agrupar por `customer_id`, no hay customer_id duplicados.<br>
Inconcluso a que se refiere con la columna `customer_unique_id` ya que son todos únicos.

### Columna `customer_zip_code_prefix`

In [132]:
# Verifico el tipo de dato este correcto
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [133]:
# Por el tipo de dato de la columna `customer_zip_code_prefix` se verifica que son todos números enteros, procedo a listarlos y contarlos.
zip_code_prefix = customers['customer_zip_code_prefix'].unique().tolist()
zip_code_prefix.sort()
len(zip_code_prefix)

14994

Verifico que los prefijos y las abreviaturas esten bien

In [134]:
# Primero verifico que los valores mínimos esten dentro del rango, luego verifico los máximos
df = customers[['customer_state','customer_zip_code_prefix']].groupby(['customer_state'], ).min()
df['prefix_max'] = customers[['customer_state','customer_zip_code_prefix']].groupby(['customer_state'], ).max().values
df.sort_values('customer_zip_code_prefix')

customer_zip_code_prefix  prefix_max
customer_state                                      
SP                                  1003       19970
RJ                                 20010       28999
ES                                 29010       29985
MG                                 30110       39995
BA                                 40010       48990
SE                                 49000       49995
PE                                 50010       56980
AL                                 57010       57975
PB                                 58010       58995
RN                                 59010       59995
CE                                 60010       63960
PI                                 64000       64980
MA                                 65010       65980
PA                                 66010       68820
AP                                 68900       68997
AM                                 69005       69800
RR                                 69301       69380
AC                                 69900       69980
DF                                 70040       73402
GO                                 72800       76730
RO                                 76801       76997
TO                                 77001       77995
MT                                 78005       78890
MS                                 79002       79995
PR                                 80010       87990
SC                                 88010       89998
RS                                 90010       99990

In [135]:
# Verifico el nombre de los estados en los intervalos para chequear
customers[(customers['customer_zip_code_prefix']>69300) & 
            (customers['customer_zip_code_prefix']<69380) & 
            (customers['customer_state']=='AM')]

Empty DataFrame
Columns: [customer_id, customer_unique_id, customer_zip_code_prefix, customer_city, customer_state]
Index: []

In [136]:
# No se pisan, ahora verifico los que estan fuera del rango de RR y siguien siendo AM
customers[(customers['customer_zip_code_prefix']>69380) & (customers['customer_state']=='AM')]

customer_id                customer_unique_id  \
5713   6e8a9c87724e18b7e567193cdd3e4b45  8fe805d171f920bf0ae393368aad5316   
56845  df6d93efba53a613bf2365f7945febb0  9484cca8fd716e613bc1a0bb17ed0172   
68437  19faaa8953bbd5166298b6f2a3f84298  3246cc1ffb8b0e719e5ba4db0443185d   
76259  a685e039c2149f0278def380b654b0b8  ccb89ab59ba8eec8518110077868630a   
80440  ec9dfbd6b5010c0bb5164e740e002d6d  e084d4891d0ffe9337c6b562b2df42b2   

       customer_zip_code_prefix customer_city customer_state  
5713                      69800       humaita             AM  
56845                     69800       humaita             AM  
68437                     69460         coari             AM  
76259                     69800       humaita             AM  
80440                     69800       humaita             AM

Los datos son correctos, ante cualquier duda referir:<br>
• humaita https://codigo-postal.org/brasil/am/humaita/<br>
• coari https://codigo-postal.org/brasil/am/coari/

In [137]:
customers[(customers['customer_zip_code_prefix']>72799) &
         (customers['customer_zip_code_prefix']<73403) & 
         (customers['customer_state']=='DF')].groupby(['customer_city'])[['customer_state']].count()

customer_state
customer_city                
brasilia                  173
sobradinho                  1

In [138]:
customers[(customers['customer_zip_code_prefix']>72799) &
         (customers['customer_zip_code_prefix']<73403) & 
         (customers['customer_state']=='DF')]

customer_id                customer_unique_id  \
144    d6ea00d4a2dca6a01c708931b325a9e5  1d4626b197f66aa6129c135ab09f1a7b   
324    8bbcf9a0717a1cb212b45522ef61972c  3078dfe83b099ae6ebe055be3abe0c0d   
1024   a55002f0d9ba4aa981a3cc93c669617a  4f463757edb0250ec353ab6f3610cf34   
1235   7c6c12489e36aafa4ab5dfcfef887475  55133cf96f44c834ac51d0e3b678c7a8   
2095   baca33004aa726524d5a891853100559  aa67457bb07fde4e952d67b038135bee   
...                                 ...                               ...   
96886  d43a38a1b87c85d51d3c00ab2dd19390  e906b389ad2bfb4058e80f63e73b3458   
97721  484572a7bd433021fd437bc6f29c8544  af7bb09e58dadca0e5f11e1a3ad0c2a9   
97986  5e28718800958c9b17749226df1f8b02  2e7bb3aad177a6e3bf8abd2de27c2a09   
98371  e62cf9e803aee2389d750a99f8a2986e  e5f6bf6e4999102bc5405eb858869122   
99367  a072679ac4ac4d56f1aaff9e763836ed  d632eddd34f00aa9d19613ba9cef6294   

       customer_zip_code_prefix customer_city customer_state  
144                       73020      brasilia             DF  
324                       73035      brasilia             DF  
1024                      73330      brasilia             DF  
1235                      73035      brasilia             DF  
2095                      73369      brasilia             DF  
...                         ...           ...            ...  
96886                     73252      brasilia             DF  
97721                     73350      brasilia             DF  
97986                     73368      brasilia             DF  
98371                     73370      brasilia             DF  
99367                     73060      brasilia             DF  

[174 rows x 5 columns]

Los datos son correctos, ante cualquier duda referir:<br>
• Brasilia https://www.worldpostalcodes.org/l1/pt/br/brasil/lista/r1/lista-dos-codigos-postais-em-distrito-federal


In [139]:
customers[(customers['customer_zip_code_prefix']>72799) &
         (customers['customer_zip_code_prefix']<73403) & 
         (customers['customer_city']=='sobradinho')]

customer_id                customer_unique_id  \
8262  075ef61489ef7122c2a0f69e79c55ff3  757423381d83adecca877fc69cfb8540   

      customer_zip_code_prefix customer_city customer_state  
8262                     73060    sobradinho             DF

Este código postal corresponde a Brasilia, que si se encuentra en Distrito Federal como indica en la siguiente columna,<br>
el correspondiente a `sobradinho` es 48925, asi que se corregirá la ciudad a Brasilia.

In [140]:
customers.iloc[8262,3]='brasilia'
customers.iloc[8262,3]

'brasilia'

### Columna `customer_city`
No vamos a trabajar con esta columna, asi que no se le hará ningún tratamiento

### Columna `customer_state`

In [141]:
print('Cantidad de federaciones:',len(customers.customer_state.unique().tolist()))
print(customers.customer_state.unique().tolist())

Cantidad de federaciones: 27
['SP', 'SC', 'MG', 'PR', 'RJ', 'RS', 'PA', 'GO', 'ES', 'BA', 'MA', 'MS', 'CE', 'DF', 'RN', 'PE', 'MT', 'AM', 'AP', 'AL', 'RO', 'PB', 'TO', 'PI', 'AC', 'SE', 'RR']


In [142]:
# Exportamos dataset limpio
customers.to_csv('./Datasets_procesados/customers.csv', index=False)

## orders

In [143]:
# Ingestamos dataset
url = 'https://raw.githubusercontent.com/ronalcabrera/PG_Olist/main/Datasets/olist_orders_dataset.csv'
orders = pd.read_csv(url, delimiter=',', encoding='UTF-8')

In [144]:
# Obervamos las primeras 15 filas del dataset
orders.head(10)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
5  a4591c265e18cb1dcee52889e2d8acc3  503740e9ca751ccdda7ba28e9ab8f608   
6  136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
7  6514b8ad8028c9f2cc2374ded245783f  9bdf08b4b3b52b5526ff42d37d47f222   
8  76c6e866289321a7c93b82b54852dc33  f54a9f0e6b351c431402b8461ea51999   
9  e69bfb5eb88e0ed6a785585b27e16dbf  31ad1d1b63eb9962463f764d4e6e0c9d   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
5    delivered      2017-07-09 21:57:05  2017-07-09 22:10:13   
6     invoiced      2017-04-11 12:22:08  2017-04-13 13:25:17   
7    delivered      2017-05-16 13:10:30  2017-05-16 13:22:11   
8    delivered      2017-01-23 18:29:09  2017-01-25 02:50:47   
9    delivered      2017-07-29 11:55:02  2017-07-29 12:05:32   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   
5          2017-07-11 14:58:04           2017-07-26 10:57:55   
6                          NaN                           NaN   
7          2017-05-22 10:07:46           2017-05-26 12:55:51   
8          2017-01-26 14:16:31           2017-02-02 14:08:10   
9          2017-08-10 19:45:24           2017-08-16 17:14:30   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00  
5           2017-08-01 00:00:00  
6           2017-05-09 00:00:00  
7           2017-06-07 00:00:00  
8           2017-03-06 00:00:00  
9           2017-08-23 00:00:00

In [145]:
# Analizamos nombres de columnas, cantidades y tipos de datos
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [146]:
# Cambiamos los tipos de datos para las fehcas
orders['order_purchase_timestamp'] = pd.to_datetime(orders['order_purchase_timestamp'], format='%Y/%m/%d')
orders['order_approved_at'] = pd.to_datetime(orders['order_approved_at'], format='%Y/%m/%d')
orders['order_delivered_carrier_date'] = pd.to_datetime(orders['order_delivered_carrier_date'], format='%Y/%m/%d')
orders['order_delivered_customer_date'] = pd.to_datetime(orders['order_delivered_customer_date'], format='%Y/%m/%d')
orders['order_estimated_delivery_date'] = pd.to_datetime(orders['order_estimated_delivery_date'], format='%Y/%m/%d')

In [147]:
# Revisamos que se haya efectuado los cambiosen el tipo de dato
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
dtypes: datetime64[ns](5), object(3)
memory usage: 6.1+ MB


In [148]:
orders.groupby(['order_status'])['order_status'].count()

order_status
approved           2
canceled         625
created            5
delivered      96478
invoiced         314
processing       301
shipped         1107
unavailable      609
Name: order_status, dtype: int64

Ahora, buscamos saber la cantidad de valores únicos, o si se repiten valores que a simple vista parezcan incoherentes.

In [149]:
# Búsqueda de valores únicos
orders.nunique()

order_id                         99441
customer_id                      99441
order_status                         8
order_purchase_timestamp         98875
order_approved_at                90733
order_delivered_carrier_date     81018
order_delivered_customer_date    95664
order_estimated_delivery_date      459
dtype: int64

In [150]:
# Exportamos dataset limpio
orders.to_csv('Datasets_procesados/orders.csv', index=False)

## geolocation

In [151]:
# Ingestamos dataset
url = 'https://raw.githubusercontent.com/ronalcabrera/PG_Olist/main/Datasets/olist_geolocation_dataset.csv'
geolocation = pd.read_csv(url, delimiter=',', encoding='UTF-8')

In [152]:
# Obervamos las primeras 15 filas del dataset
geolocation.head(15)

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                          1037       -23.545621       -46.639292   
1                          1046       -23.546081       -46.644820   
2                          1046       -23.546129       -46.642951   
3                          1041       -23.544392       -46.639499   
4                          1035       -23.541578       -46.641607   
5                          1012       -23.547762       -46.635361   
6                          1047       -23.546273       -46.641225   
7                          1013       -23.546923       -46.634264   
8                          1029       -23.543769       -46.634278   
9                          1011       -23.547640       -46.636032   
10                         1013       -23.547325       -46.634184   
11                         1032       -23.538418       -46.634778   
12                         1014       -23.546435       -46.633830   
13                         1012       -23.548946       -46.634671   
14                         1037       -23.545187       -46.637855   

   geolocation_city geolocation_state  
0         sao paulo                SP  
1         sao paulo                SP  
2         sao paulo                SP  
3         sao paulo                SP  
4         sao paulo                SP  
5         são paulo                SP  
6         sao paulo                SP  
7         sao paulo                SP  
8         sao paulo                SP  
9         sao paulo                SP  
10        sao paulo                SP  
11        sao paulo                SP  
12        sao paulo                SP  
13        sao paulo                SP  
14        são paulo                SP

In [153]:
# Analizamos nombres de columnas, cantidades y tipos de datos
geolocation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


Ahora, buscamos saber la cantidad de valores únicos, o si se repiten valores que a simple vista parezcan incoherentes.

In [154]:
# Búsqueda de valores únicos
geolocation.nunique()

geolocation_zip_code_prefix     19015
geolocation_lat                717360
geolocation_lng                717613
geolocation_city                 8011
geolocation_state                  27
dtype: int64

Hay 19015 zip_code únicos, frente a 1000163 registros. Hay que reducir la tabla respecto de la cantidad de zip_code únicos

In [155]:
geolocation.duplicated().value_counts()

False    738332
True     261831
dtype: int64

Exploro los nombres y orden de columnas

In [156]:
geolocation.columns

Index(['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng',
       'geolocation_city', 'geolocation_state'],
      dtype='object')

In [157]:
# Exportamos dataset limpio
geolocation.to_csv('Datasets_Procesados/geolocation.csv', index=False)

In [158]:
import pandas as pd
import numpy as np

## marketing_qualified_leads

In [159]:
# Ingestamos dataset
url='https://raw.githubusercontent.com/ronalcabrera/PG_Olist/main/Datasets/olist_marketing_qualified_leads_dataset.csv'
marketing_qualified_leads=pd.read_csv(url, delimiter=',', encoding='UTF-8')

In [160]:
marketing_qualified_leads

mql_id first_contact_date  \
0     dac32acd4db4c29c230538b72f8dd87d         2018-02-01   
1     8c18d1de7f67e60dbd64e3c07d7e9d5d         2017-10-20   
2     b4bc852d233dfefc5131f593b538befa         2018-03-22   
3     6be030b81c75970747525b843c1ef4f8         2018-01-22   
4     5420aad7fec3549a85876ba1c529bd84         2018-02-21   
...                                ...                ...   
7995  feaba3ffcd2ff97501696c7f9a42f41c         2018-05-22   
7996  a79cb53cd009ab92e0143b92baa2407b         2018-03-27   
7997  68f049a23ab109c6a0f6989bb9a02994         2017-08-27   
7998  4f8c96e2509b984329044c6682c88ee9         2017-10-06   
7999  8a6492305a5fbcdcdd1a7f5a90764c07         2018-01-11   

                       landing_page_id          origin  
0     88740e65d5d6b056e0cda098e1ea6313          social  
1     007f9098284a86ee80ddeb25d53e0af8     paid_search  
2     a7982125ff7aa3b2054c6e44f9d28522  organic_search  
3     d45d558f0daeecf3cccdffe3c59684aa           email  
4     b48ec5f3b04e9068441002a19df93c6c  organic_search  
...                                ...             ...  
7995  e42a14209c69c3e9cc6b042620465f12     paid_search  
7996  c494978688ccf66ad9fad3d6a3338c22     paid_search  
7997  b48ec5f3b04e9068441002a19df93c6c  organic_search  
7998  a56671a54260a44923d32c2f08fad39c  organic_search  
7999  b76ef37428e6799c421989521c0e5077  direct_traffic  

[8000 rows x 4 columns]

In [161]:
marketing_qualified_leads.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   mql_id              8000 non-null   object
 1   first_contact_date  8000 non-null   object
 2   landing_page_id     8000 non-null   object
 3   origin              7940 non-null   object
dtypes: object(4)
memory usage: 250.1+ KB


In [162]:
# En la columna 'origin' reemplamos los NaN por 'Sin Dato'
marketing_qualified_leads['origin']=marketing_qualified_leads['origin'].replace(np.nan,'Sin Dato')

In [163]:
marketing_qualified_leads.isnull().sum()

mql_id                0
first_contact_date    0
landing_page_id       0
origin                0
dtype: int64

In [164]:
# Cambiaremos el tipo de dato de la columna 'first_contact_date'
marketing_qualified_leads['first_contact_date']=marketing_qualified_leads['first_contact_date'].astype('datetime64')

In [165]:
# Exportamos dataset limpio
marketing_qualified_leads.to_csv('./Datasets_procesados/marketing_qualified_leads.csv', index=False)

## order_items

In [166]:
# Ingestamos dataset
url = 'https://raw.githubusercontent.com/ronalcabrera/PG_Olist/main/Datasets/olist_order_items_dataset.csv'
df_order_items = pd.read_csv(url, delimiter=',', encoding='UTF-8')

In [167]:
df_order_items.sample(5)

order_id  order_item_id  \
112068  fe99f8be754e8a75c4201c7156bd99ee              1   
16847   2695fad29683cb9e20b91e87fc0cfd39              1   
19283   2c494cef88b908df152b7d6d69c8beba              1   
69295   9e40d6ad0e658dbe287498d89f222d18              1   
98096   de852e80c0b14138d06937c2695b4b94              1   

                              product_id                         seller_id  \
112068  56ad64ea1d19bc4b26b61088d89e6f41  30a2f535bb48308f991d0b9ad4a8c4bb   
16847   dab2413ead0edda9967edbc9bda2a64e  ea8482cd71df3c1969d7b9473ff13abc   
19283   1405f399d2e4807af6ce6a2519e2ed7b  973f21788dfab357250f69a8dcb7ddee   
69295   6dbfe807ef0967e3a7136dee5bbe90ed  93dc87703c046b603023e75222018b45   
98096   38d070104c7e27171e177d581758a90f  fb89df7f89fd1b8f03fa2acca1ba7364   

        shipping_limit_date   price  freight_value  
112068  2017-06-08 14:43:00   49.90          17.63  
16847   2018-04-17 15:35:18   29.98           7.39  
19283   2018-05-17 11:37:28  105.00          41.68  
69295   2018-06-14 16:12:19  179.00          16.13  
98096   2018-07-11 20:50:50   37.00          18.36

In [168]:
df_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [169]:
# Verificamos cantidad de nulos por columna (no hay nulos)

df_order_items.isnull().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [170]:
df_order_items.drop_duplicates() # No hay duplicados

order_id  order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214              1   
1       00018f77f2f0320c557190d7a144bdd3              1   
2       000229ec398224ef6ca0657da4fc703e              1   
3       00024acbcdf0a6daa1e931b038114c75              1   
4       00042b26cf59d7ce69dfabb4e55b4fd9              1   
...                                  ...            ...   
112645  fffc94f6ce00a00581880bf54a75a037              1   
112646  fffcd46ef2263f404302a634eb57f7eb              1   
112647  fffce4705a9662cd70adb13d4a31832d              1   
112648  fffe18544ffabc95dfada21779c9644f              1   
112649  fffe41c64501cc87c801fd61db3f6244              1   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2       c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3       7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4       ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
...                                  ...                               ...   
112645  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
112646  32e07fd915822b0765e448c4dd74c828  f3c38ab652836d21de61fb8314b69182   
112647  72a30483855e2eafc67aee5dc2560482  c3cfdc648177fdbbbb35635a37472c53   
112648  9c422a519119dcad7575db5af1ba540e  2b3e4a2a3ea8e01938cabda2a3e5cc79   
112649  350688d9dc1e75ff97be326363655e01  f7ccf836d21b2fb1de37564105216cc1   

        shipping_limit_date   price  freight_value  
0       2017-09-19 09:45:35   58.90          13.29  
1       2017-05-03 11:05:13  239.90          19.93  
2       2018-01-18 14:48:30  199.00          17.87  
3       2018-08-15 10:10:18   12.99          12.79  
4       2017-02-13 13:57:51  199.90          18.14  
...                     ...     ...            ...  
112645  2018-05-02 04:11:01  299.99          43.41  
112646  2018-07-20 04:31:48  350.00          36.53  
112647  2017-10-30 17:14:25   99.90          16.95  
112648  2017-08-21 00:04:32   55.99           8.72  
112649  2018-06-12 17:10:13   43.00          12.79  

[112650 rows x 7 columns]

In [171]:
# Cambiar la columna shipping_limit_date a tipo datetime
df_order_items['shipping_limit_date'] = pd.to_datetime(df_order_items['shipping_limit_date'])

In [172]:
df_order_items.sample(5)

order_id  order_item_id  \
36296  5251ac9886fbf2953b24c9f45364837e              1   
65546  95d825d3c9ac0db6ca8f1e567cec7586              1   
37615  5559edd02dabfec0a6ee808960297d1b              1   
52631  77c374d0a559e948d26ddd439b558558              1   
80989  b822c2238bcf6e7dac25f9cba3312a02              1   

                             product_id                         seller_id  \
36296  39084b68d30ae3ce2bdaf467ed4f3dbc  5f1dc28029d2c244352a68107ec2b542   
65546  3ec2984cb2de1675a0306bfc60dbe5a5  0c8380b62e38e8a1e6adbeba7eb9688c   
37615  4b96786612ebe7463132fce2c4dca136  d94a40fd42351c259927028d163af842   
52631  f264c1d9b20b5e4a340254d0405e613b  7a67c85e85bb2ce8582c35f2203ad736   
80989  5215505db2113a032810c3644c346a38  6560211a19b47992c3666cc44a7e94c0   

      shipping_limit_date   price  freight_value  
36296 2017-09-25 16:43:26   39.50          15.10  
65546 2017-03-20 16:47:29  119.90          19.12  
37615 2017-11-30 20:52:32   83.30          21.07  
52631 2017-12-20 08:52:15   72.99          13.53  
80989 2018-03-06 22:20:22   99.00           7.95

In [173]:
df_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  object        
 1   order_item_id        112650 non-null  int64         
 2   product_id           112650 non-null  object        
 3   seller_id            112650 non-null  object        
 4   shipping_limit_date  112650 non-null  datetime64[ns]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(3)
memory usage: 6.0+ MB


In [174]:
# Exportamos dataset limpio
df_order_items.to_csv('./Datasets_procesados/order_items.csv', index=False)

## order_payments

In [175]:
# Ingestamos dataset
url='https://raw.githubusercontent.com/ronalcabrera/PG_Olist/main/Datasets/olist_order_payments_dataset.csv'
payments= pd.read_csv(url, delimiter=',', encoding='UTF-8')

In [176]:
payments

order_id  payment_sequential payment_type  \
0       b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1       a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2       25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3       ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4       42fdf880ba16b47b59251dd489d4441a                   1  credit_card   
...                                  ...                 ...          ...   
103881  0406037ad97740d563a178ecc7a2075c                   1       boleto   
103882  7b905861d7c825891d6347454ea7863f                   1  credit_card   
103883  32609bbb3dd69b3c066a6860554a77bf                   1  credit_card   
103884  b8b61059626efa996a60be9bb9320e10                   1  credit_card   
103885  28bbae6599b09d39ca406b747b6632b1                   1       boleto   

        payment_installments  payment_value  
0                          8          99.33  
1                          1          24.39  
2                          1          65.71  
3                          8         107.78  
4                          2         128.45  
...                      ...            ...  
103881                     1         363.31  
103882                     2          96.80  
103883                     1          47.77  
103884                     5         369.54  
103885                     1         191.58  

[103886 rows x 5 columns]

In [177]:
payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [178]:
# Exportamos dataset limpio
payments.to_csv('./Datasets_procesados/order_payments.csv', index=False)

## order_reviews

In [179]:
# Ingestamos dataset
url='https://raw.githubusercontent.com/ronalcabrera/PG_Olist/main/Datasets/olist_order_reviews_dataset.csv'
reviews=pd.read_csv(url, delimiter=',', encoding='UTF-8')

In [180]:
reviews

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4                  NaN   
1                 5                  NaN   
2                 5                  NaN   
3                 5                  NaN   
4                 5                  NaN   
...             ...                  ...   
99219             5                  NaN   
99220             5                  NaN   
99221             5                  NaN   
99222             4                  NaN   
99223             1                  NaN   

                                  review_comment_message review_creation_date  \
0                                                    NaN  2018-01-18 00:00:00   
1                                                    NaN  2018-03-10 00:00:00   
2                                                    NaN  2018-02-17 00:00:00   
3                  Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4      Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
...                                                  ...                  ...   
99219                                                NaN  2018-07-07 00:00:00   
99220                                                NaN  2017-12-09 00:00:00   
99221  Excelente mochila, entrega super rápida. Super...  2018-03-22 00:00:00   
99222                                                NaN  2018-07-01 00:00:00   
99223  meu produto chegou e ja tenho que devolver, po...  2017-07-03 00:00:00   

      review_answer_timestamp  
0         2018-01-18 21:46:59  
1         2018-03-11 03:05:13  
2         2018-02-18 14:36:24  
3         2017-04-21 22:02:06  
4         2018-03-02 10:26:53  
...                       ...  
99219     2018-07-14 17:18:30  
99220     2017-12-11 20:06:42  
99221     2018-03-23 09:10:43  
99222     2018-07-02 12:59:13  
99223     2017-07-03 21:01:49  

[99224 rows x 7 columns]

In [181]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [182]:
#Casos a tener en cuenta
# En la columna review_comment_title tiene un número en lugar de un título.
#reviews[reviews['review_id']=='14613df954694f243dedeaa2c8a3a5c2']

#En la columna review_comment_message aparecen fechas //  o ///
#reviews[(reviews['review_id']=="7d9de731a6322fddaffbb1626413466a") | (reviews['review_id']=="1f3b2abc10c150806a54ddb7ad99f8de")]

#En la columna review_comment_message un mensaje //// //////... :
#reviews[reviews['review_id']=="7915e960f5094fc36e0103b1592f92ba"]

##### Columna `review_comment_title`


In [183]:
# Reemplazo los NaN por 'Sin Dato'
reviews['review_comment_title'] = reviews['review_comment_title'].replace(np.nan,'Sin Dato')
# Reemplazo los saltos de lineas por ' ' 
reviews['review_comment_title'] = reviews['review_comment_title'].str.replace('\n',' ')
# Remplezo en los mensajes que inician con el o finalizan con "
reviews['review_comment_title'] = reviews['review_comment_title'].str.replace('"','')

##### Columna `review_comment_message`

In [184]:
# Reemplazo los NaN por 'Sin Dato'
reviews['review_comment_message'] = reviews['review_comment_message'].replace(np.nan,'Sin Dato')
# Reemplazo los saltos de linea por ' '
reviews['review_comment_message'] = reviews['review_comment_message'].str.replace('\n',' ')
#
reviews['review_comment_message'] = reviews['review_comment_message'].str.replace('"','')

In [185]:
reviews.isnull().sum()

review_id                  0
order_id                   0
review_score               0
review_comment_title       0
review_comment_message     0
review_creation_date       0
review_answer_timestamp    0
dtype: int64

##### Columna `review_creation_date`

In [186]:
# Al hacer este cambio de tipo de dato, la hora se sacará porque todas son 00

# Cambio tipo de dato : de 'object' a 'datetime64'
reviews['review_creation_date']=reviews['review_creation_date'].astype('datetime64')

##### Columna `review_answer_timestamp`

In [187]:
# Cambio tipo de dato : de 'object' a 'datetime64'
reviews['review_answer_timestamp']=reviews['review_answer_timestamp'].astype('datetime64')

In [188]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                99224 non-null  object        
 1   order_id                 99224 non-null  object        
 2   review_score             99224 non-null  int64         
 3   review_comment_title     99224 non-null  object        
 4   review_comment_message   99224 non-null  object        
 5   review_creation_date     99224 non-null  datetime64[ns]
 6   review_answer_timestamp  99224 non-null  datetime64[ns]
dtypes: datetime64[ns](2), int64(1), object(4)
memory usage: 5.3+ MB


In [189]:
# Exportamos dataset limpio
reviews.to_csv('./Datasets_procesados/order_reviews.csv', index=False)

## closed_deals

In [190]:
# Ingestamos dataset
url = 'https://raw.githubusercontent.com/ronalcabrera/PG_Olist/main/Datasets/olist_closed_deals_dataset.csv'
df_closed_deals = pd.read_csv(url, delimiter=',', encoding='UTF-8')

In [191]:
df_closed_deals.sample(5)

mql_id                         seller_id  \
465  b2fb7865dfca2461177cbad7ec520b0e  e1c76f339ebd5460999f25a2aa8c92c5   
117  fc0a87e43ede1f423412583293815598  90c523952c5622d268addb6d93a7212f   
167  6c465ed53d17739467c95fe414d8e056  3289d8c3eef4873f91ce0b1343c0dd58   
227  bd0c413c4350a9a29fb9bec5de64e2a6  2aeedd049dc20e7171f43110fd0d6821   
213  cf8f0678e67cb002c2b68f9f68629885  bed6d1f6c4664ab8cf3abf3821727756   

                               sdr_id                             sr_id  \
465  56bf83c4bb35763a51c2baab501b4c67  c638112b43f1d1b86dcabb0da720c901   
117  f42a2bd194f7802ab052a815c8de65b7  d3d1e91a157ea7f90548eef82f1955e3   
167  a8387c01a09e99ce014107505b92388c  4ef15afb4b2723d8f3d81e51ec7afefe   
227  9d12ef1a7eca3ec58c545c678af7869c  495d4e95a8cf8bbf8b432b612a2aa328   
213  068066e24f0c643eb1d089c7dd20cd73  85fc447d336637ba1df43e793199fbc8   

                won_date                 business_segment        lead_type  \
465  2018-02-08 17:10:00                 small_appliances    online_medium   
117  2018-08-20 13:59:51          audio_video_electronics  online_beginner   
167  2018-04-26 20:04:29                   sports_leisure     online_small   
227  2018-06-13 16:37:42  construction_tools_house_garden         industry   
213  2018-03-22 21:20:42  construction_tools_house_garden         industry   

    lead_behaviour_profile has_company has_gtin average_stock business_type  \
465                    NaN         NaN      NaN           NaN           NaN   
117                    cat        True     True           NaN      reseller   
167                    cat         NaN      NaN           NaN  manufacturer   
227                    cat         NaN      NaN           NaN      reseller   
213                  shark         NaN      NaN           NaN      reseller   

     declared_product_catalog_size  declared_monthly_revenue  
465                            NaN                       0.0  
117                            5.0                       0.0  
167                            NaN                       0.0  
227                            NaN                       0.0  
213                            NaN                       0.0

In [192]:
# Tipo de datos de cada columna 
df_closed_deals.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 842 entries, 0 to 841
Data columns (total 14 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   mql_id                         842 non-null    object 
 1   seller_id                      842 non-null    object 
 2   sdr_id                         842 non-null    object 
 3   sr_id                          842 non-null    object 
 4   won_date                       842 non-null    object 
 5   business_segment               841 non-null    object 
 6   lead_type                      836 non-null    object 
 7   lead_behaviour_profile         665 non-null    object 
 8   has_company                    63 non-null     object 
 9   has_gtin                       64 non-null     object 
 10  average_stock                  66 non-null     object 
 11  business_type                  832 non-null    object 
 12  declared_product_catalog_size  69 non-null     flo

In [193]:
# Verificamos cantidad de nulos por columna
df_closed_deals.isnull().sum()

mql_id                             0
seller_id                          0
sdr_id                             0
sr_id                              0
won_date                           0
business_segment                   1
lead_type                          6
lead_behaviour_profile           177
has_company                      779
has_gtin                         778
average_stock                    776
business_type                     10
declared_product_catalog_size    773
declared_monthly_revenue           0
dtype: int64

In [194]:
df_closed_deals.drop_duplicates() # No hay duplicados

mql_id                         seller_id  \
0    5420aad7fec3549a85876ba1c529bd84  2c43fb513632d29b3b58df74816f1b06   
1    a555fb36b9368110ede0f043dfc3b9a0  bbb7d7893a450660432ea6652310ebb7   
2    327174d3648a2d047e8940d7d15204ca  612170e34b97004b3ba37eae81836b4c   
3    f5fee8f7da74f4887f5bcae2bafb6dd6  21e1781e36faf92725dde4730a88ca0f   
4    ffe640179b554e295c167a2f6be528e0  ed8cb7b190ceb6067227478e48cf8dde   
..                                ...                               ...   
837  df9a2abe2ad3f2d8758b675ac8331ecf  bfcc27719640628da877db48b672b169   
838  8723201570415eb23b4f095d8c20a27c  4a82eab98441aeb64566e2776c1fb2b6   
839  2e57665b8faf05c967a801eb5aedfa0a  55031883943971ca22db6894574cfe2c   
840  40955be51cc85c5a5a6cdd12b19e9f10  c0e933c238e41f0cd459d6025ee9b364   
841  8a6492305a5fbcdcdd1a7f5a90764c07  1bba9c94eb5c4e82fba9f7286fe81025   

                               sdr_id                             sr_id  \
0    a8387c01a09e99ce014107505b92388c  4ef15afb4b2723d8f3d81e51ec7afefe   
1    09285259593c61296eef10c734121d5b  d3d1e91a157ea7f90548eef82f1955e3   
2    b90f87164b5f8c2cfa5c8572834dbe3f  6565aa9ce3178a5caf6171827af3a9ba   
3    56bf83c4bb35763a51c2baab501b4c67  d3d1e91a157ea7f90548eef82f1955e3   
4    4b339f9567d060bcea4f5136b9f5949e  d3d1e91a157ea7f90548eef82f1955e3   
..                                ...                               ...   
837  45749fb708130f78d0db07d8d80f030b  a8387c01a09e99ce014107505b92388c   
838  370c9f455f93a9a96cbe9bea48e70033  4b339f9567d060bcea4f5136b9f5949e   
839  370c9f455f93a9a96cbe9bea48e70033  a8387c01a09e99ce014107505b92388c   
840  068066e24f0c643eb1d089c7dd20cd73  d3d1e91a157ea7f90548eef82f1955e3   
841  56bf83c4bb35763a51c2baab501b4c67  de63de0d10a6012430098db33c679b0b   

                won_date                 business_segment        lead_type  \
0    2018-02-26 19:58:54                              pet    online_medium   
1    2018-05-08 20:17:59                  car_accessories         industry   
2    2018-06-05 17:27:23                  home_appliances       online_big   
3    2018-01-17 13:51:03                       food_drink     online_small   
4    2018-07-03 20:17:45                  home_appliances         industry   
..                   ...                              ...              ...   
837  2018-08-01 20:22:31                       home_decor          offline   
838  2018-09-27 18:58:41  construction_tools_house_garden       online_big   
839  2018-06-06 20:15:26                        computers    online_medium   
840  2018-03-06 15:52:23                              pet  online_beginner   
841  2018-05-02 13:31:32              household_utilities         industry   

    lead_behaviour_profile has_company has_gtin average_stock business_type  \
0                      cat         NaN      NaN           NaN      reseller   
1                    eagle         NaN      NaN           NaN      reseller   
2                      cat         NaN      NaN           NaN      reseller   
3                      NaN         NaN      NaN           NaN      reseller   
4                     wolf         NaN      NaN           NaN  manufacturer   
..                     ...         ...      ...           ...           ...   
837                    cat         NaN      NaN           NaN      reseller   
838                   wolf         NaN      NaN           NaN      reseller   
839                   wolf         NaN      NaN           NaN      reseller   
840                    cat         NaN      NaN           NaN      reseller   
841                  eagle         NaN      NaN           NaN  manufacturer   

     declared_product_catalog_size  declared_monthly_revenue  
0                              NaN                       0.0  
1                              NaN                       0.0  
2                              NaN                       0.0  
3                              NaN                       0.0  
4                           

In [195]:
# Cambiar la columna won_date a tipo datetime
df_closed_deals['won_date'] = pd.to_datetime(df_closed_deals['won_date'])

In [196]:
df_closed_deals.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 842 entries, 0 to 841
Data columns (total 14 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   mql_id                         842 non-null    object        
 1   seller_id                      842 non-null    object        
 2   sdr_id                         842 non-null    object        
 3   sr_id                          842 non-null    object        
 4   won_date                       842 non-null    datetime64[ns]
 5   business_segment               841 non-null    object        
 6   lead_type                      836 non-null    object        
 7   lead_behaviour_profile         665 non-null    object        
 8   has_company                    63 non-null     object        
 9   has_gtin                       64 non-null     object        
 10  average_stock                  66 non-null     object        
 11  business_type      

In [197]:
# Verificamos valores unicos para encontrar redacciones erróneas y repararlas

print(df_closed_deals["business_segment"].unique())
print(df_closed_deals["lead_type"].unique())
print(df_closed_deals["lead_behaviour_profile"].unique())
print(df_closed_deals["has_company"].unique())
print(df_closed_deals["has_gtin"].unique())
print(df_closed_deals["average_stock"].unique())
print(df_closed_deals["business_type"].unique())

['pet' 'car_accessories' 'home_appliances' 'food_drink' 'health_beauty'
 'computers' 'household_utilities' 'construction_tools_house_garden'
 'toys' 'sports_leisure' 'stationery' 'food_supplement' 'home_decor'
 'bed_bath_table' 'watches' 'fashion_accessories' 'jewerly' 'party'
 'small_appliances' 'audio_video_electronics' 'other' 'bags_backpacks'
 'home_office_furniture' 'music_instruments' 'books' 'baby'
 'air_conditioning' 'phone_mobile' nan 'handcrafted' 'perfume' 'gifts'
 'religious' 'games_consoles']
['online_medium' 'industry' 'online_big' 'online_small' 'offline'
 'online_top' 'online_beginner' 'other' nan]
['cat' 'eagle' nan 'wolf' 'shark' 'cat, wolf' 'eagle, wolf' 'shark, cat'
 'eagle, cat' 'shark, wolf']
[nan True False]
[nan True False]
[nan '20-50' '1-5' '5-20' '200+' '50-200' 'unknown']
['reseller' 'manufacturer' 'other' nan]


In [198]:
# Reemplazamos los valores NaN por 'Sin dato'
df_closed_deals.fillna({'business_segment': 'Sin dato', 'lead_type': 'Sin dato', 'lead_behaviour_profile': 'Sin dato', 
                        'has_company': 'Sin dato' ,'has_gtin': 'Sin dato','average_stock': 'Sin dato',
                        'business_type': 'Sin dato'}, inplace=True)

In [199]:
# Reemplazamos en average_stock los 'unknown' por 'Sin dato'
df_closed_deals['average_stock'].replace('unknown','Sin dato', inplace=True)

In [200]:
# OUTLIERS
#ventas.loc[ventas['declared_monthly_revenue'] < 1000,'declared_monthly_revenue'] = np.nan
#ventas.loc[ventas['declared_monthly_revenue'] > 500000,'declared_monthly_revenue'] = np.nan

In [201]:
df_closed_deals.isnull().sum()

mql_id                             0
seller_id                          0
sdr_id                             0
sr_id                              0
won_date                           0
business_segment                   0
lead_type                          0
lead_behaviour_profile             0
has_company                        0
has_gtin                           0
average_stock                      0
business_type                      0
declared_product_catalog_size    773
declared_monthly_revenue           0
dtype: int64

In [202]:
# Exportamos nuevo dataset
df_closed_deals.to_csv('./Datasets_procesados/closed_deals.csv', index=False)